app/web_conn.go

```diff
@@ -121,45 +131,95 @@ func (wc *WebConn) SetSession(v *model.Session) {
-// is ready to send/receive messages.
+// is ready to send messages.
 func (wc *WebConn) Pump() {
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		wc.writePump()
-	}()
-	wc.readPump()
-	close(wc.endWritePump)
-	wg.Wait()
+	// writePump is blocking in nature.
+	wc.writePump()
+	// Once it exits, we close everything.
 	wc.App.HubUnregister(wc)
 	close(wc.pumpFinished)
 }
 
-func (wc *WebConn) readPump() {
-	defer func() {
-		wc.WebSocket.Close()
-	}()
-	wc.WebSocket.SetReadLimit(model.SOCKET_MAX_MESSAGE_SIZE_KB)
+// startPoller adds the file descriptor of the connection
+// to the global epoll instance and registers a callback.
+func (wc *WebConn) startPoller() {
+	desc := netpoll.Must(netpoll.HandleRead(wc.WebSocket))
+	wc.App.Srv().Poller().Start(desc, func(wsEv netpoll.Event) {
+		if wsEv&(netpoll.EventReadHup|netpoll.EventHup) != 0 {
+			wc.App.Srv().Poller().Stop(desc)
+			wc.Close()
+			return
+		}
+
+		// Block until we have a token.
+		wc.App.Srv().GetWebConnToken()
+		// Read from conn.
+		go func() {
+			defer wc.App.Srv().ReleaseWebConnToken()
+			err := wc.ReadMsg()
+			if err != nil {
+				mlog.Debug("Error while reading message from websocket", mlog.Err(err))
+				wc.App.Srv().Poller().Stop(desc)
+				wc.Close()
+			}
+		}()
+	})
+}
+
+// GetWebConnToken creates backpressure by using
+// a counting semaphore to limit the number of concurrent goroutines.
+func (s *Server) GetWebConnToken() {
+	s.webConnSemaWg.Add(1)
+	s.webConnSema <- struct{}{}
+}
+
+// ReleaseWebConnToken releases a token
+// got from the semaphore
+func (s *Server) ReleaseWebConnToken() {
+	<-s.webConnSema
+	s.webConnSemaWg.Done()
+}
+
+// ReadMsg will read a single message from the websocket connection.
+func (wc *WebConn) ReadMsg() error {
 	wc.WebSocket.SetReadDeadline(time.Now().Add(pongWaitTime))
-	wc.WebSocket.SetPongHandler(func(string) error {
-		wc.WebSocket.SetReadDeadline(time.Now().Add(pongWaitTime))
+
+	r := wsutil.NewReader(wc.WebSocket, ws.StateServerSide)
+	r.MaxFrameSize = model.SOCKET_MAX_MESSAGE_SIZE_KB
+	decoder := json.NewDecoder(r)
+
+	// The reader's methods are not goroutine safe.
+	// We restrict only one reader goroutine per-connection.
+	wc.readMut.Lock()
+	defer wc.readMut.Unlock()
+
+	var req model.WebSocketRequest
+	hdr, err := r.NextFrame()
+	if err != nil {
+		return errors.Wrap(err, "error while getting the next websocket frame")
+	}
+	switch hdr.OpCode {
+	case ws.OpClose:
+		// Return if closed.
+		return nil
+	case ws.OpPong:
+		// Handle pongs
 		if wc.IsAuthenticated() {
 			wc.App.Srv().Go(func() {
 				wc.App.SetStatusAwayIfNeeded(wc.UserId, false)
 			})
 		}
-		return nil
-	})
-
-	for {
-		var req model.WebSocketRequest
-		if err := wc.WebSocket.ReadJSON(&req); err != nil {
-			wc.logSocketErr("websocket.read", err)
-			return
+	default:
+		// Default case of data message.
+		if err := decoder.Decode(&req); err != nil {
+			// We discard any remaining data left in the socket.
+			r.Discard()
```

Comment:
Do we need to do that? Is it not handled already in the receiver? since the receiver of this when it gets an error closes the poller all together.

Reply:
Ignore me, i see that this is the socket itself.